File: sdk/python/inkbox/phone/resources/hosted_agent.py
"""
inkbox/phone/resources/hosted_agent.py

Per-identity hosted call agent config: get_config, set_config.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from uuid import UUID

from inkbox.phone.types import HostedAgentConfig

if TYPE_CHECKING:
    from inkbox._http import HttpTransport


class HostedAgentConfigResource:

    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def get_config(
        self,
        *,
        agent_identity_id: UUID | str | None = None,
    ) -> HostedAgentConfig:
        """Get the hosted call agent config for an identity.

        ``agent_identity_id`` is optional — an agent-scoped key resolves
        its own identity; under admin/JWT the server 422s if it's omitted.

        Args:
            agent_identity_id: UUID of the agent identity, or ``None`` for
                an agent-scoped key.
        """
        params: dict[str, Any] = {}
        if agent_identity_id is not None:
            params["agent_identity_id"] = str(agent_identity_id)
        data = self._http.get("/hosted-agent-config", params=params)
        return HostedAgentConfig._from_dict(data)

    def set_config(
        self,
        *,
        voice: str | None = None,
        model: str | None = None,
        instructions: str | None = None,
        agent_identity_id: UUID | str | None = None,
    ) -> HostedAgentConfig:
        """Set the hosted call agent config for an identity.

        Full-replace PUT: every call sets all three fields, and a field
        left at ``None`` resets to the server default — there is no
        partial update.

        Args:
            voice: Voice override, or ``None`` for the server default.
            model: Model override, or ``None`` for the server default.
            instructions: Per-identity steering prompt appended to the
                hosted agent's system prompt, or ``None`` for none.
            agent_identity_id: UUID of the agent identity, or ``None`` for
                an agent-scoped key.
        """
        body: dict[str, Any] = {}
        if agent_identity_id is not None:
            body["agent_identity_id"] = str(agent_identity_id)
        # Omitted (None) fields are equivalent to explicit nulls on this
        # full-replace PUT: the server resets them to its defaults.
        if voice is not None:
            body["voice"] = voice
        if model is not None:
            body["model"] = model
        if instructions is not None:
            body["instructions"] = instructions
        data = self._http.put("/hosted-agent-config", json=body)
        return HostedAgentConfig._from_dict(data)

File: sdk/typescript/src/phone/resources/hostedAgent.ts
/**
 * inkbox-phone/resources/hostedAgent.ts
 *
 * Identity-scoped Inkbox Voice AI config (getConfig / setConfig).
 */

import { HttpTransport } from "../../_http.js";
import {
  HostedAgentConfig,
  RawHostedAgentConfig,
  parseHostedAgentConfig,
} from "../types.js";

export class HostedAgentConfigResource {
  constructor(private readonly http: HttpTransport) {}

  /**
   * Get the Inkbox Voice AI config.
   *
   * Agent-scoped keys resolve their own identity; admin/JWT callers must
   * pass `agentIdentityId` (the server returns 422 otherwise).
   *
   * @param options.agentIdentityId - UUID of the agent identity. Optional
   *   for agent-scoped keys; required under admin/JWT.
   */
  async getConfig(options?: {
    agentIdentityId?: string;
  }): Promise<HostedAgentConfig> {
    const params: Record<string, string> = {};
    if (options?.agentIdentityId !== undefined) {
      params["agent_identity_id"] = options.agentIdentityId;
    }
    const data = await this.http.get<RawHostedAgentConfig>(
      "/hosted-agent-config",
      params,
    );
    return parseHostedAgentConfig(data);
  }

  /**
   * Set the Inkbox Voice AI config.
   *
   * Full-replace PUT: every call sets all three fields, and a field left
   * undefined resets to the server default — there is no partial update.
   *
   * @param options.voice - Voice override; omit for the server default.
   * @param options.model - Model override; omit for the server default.
   * @param options.instructions - Per-identity steering prompt appended to
   *   Voice AI's system prompt; omit for none.
   * @param options.agentIdentityId - UUID of the agent identity. Optional
   *   for agent-scoped keys; required under admin/JWT.
   */
  async setConfig(options?: {
    voice?: string;
    model?: string;
    instructions?: string;
    agentIdentityId?: string;
  }): Promise<HostedAgentConfig> {
    const body: Record<string, unknown> = {};
    if (options?.agentIdentityId !== undefined) {
      body["agent_identity_id"] = options.agentIdentityId;
    }
    // Omitted fields are equivalent to explicit nulls on this full-replace
    // PUT: the server resets them to its defaults.
    if (options?.voice !== undefined) body["voice"] = options.voice;
    if (options?.model !== undefined) body["model"] = options.model;
    if (options?.instructions !== undefined) {
      body["instructions"] = options.instructions;
    }
    const data = await this.http.put<RawHostedAgentConfig>(
      "/hosted-agent-config",
      body,
    );
    return parseHostedAgentConfig(data);
  }
}
